step02/app.py

```diff
@@ -0,0 +1,84 @@
+from flask import Flask, jsonify, request # web framework
+import boto3 # aws SDK
+
+# openAPI spec
+from openapi_core import create_spec
+from openapi_core.contrib.flask.decorators import FlaskOpenAPIViewDecorator
+import yaml
+
+# DB
+import db
+from models import Images
+
+# misc
+import uuid
+import pathlib
+
+with open('api.yaml') as infile:
+  spec_dict = yaml.load(infile.read())
+  spec = create_spec(spec_dict)
+
+openapi = FlaskOpenAPIViewDecorator.from_spec(spec)
+path = pathlib.PurePath
+
+BUCKET_NAME = 'gg-photo-bucket'
+UPLOAD_FOLDER = 'images/'
+AWS_PROFILE = 'thumbnail-service'
+
+session = boto3.Session(profile_name=AWS_PROFILE)
+s3 = session.client('s3')
+
+app = Flask(__name__)
+
+
+@app.route('/users/me', methods=['GET'])
+@openapi
+def get_profile():
+  return jsonify({
+    'id': 0
+  })
+
+@app.route('/images/upload_url', methods=['GET'])
+@openapi
+def get_upload_url():
+    # query param
+    filename = request.openapi.parameters.query['filename']
+    content_type = request.openapi.parameters.query['content-type']
+    # values used for S3 presigned-url
+    stem = path(filename).stem
+    file_uuid = uuid.uuid4()
+    key = UPLOAD_FOLDER + f'{file_uuid}'
```

Comment:
Probably better to do something like `key=f'{UPLOAD_FOLDER}/{file_uuid}` or better yet use something like PurePath from pathlib to build the key.

In the future if we update the code so that it is configurable from an environment variable like `UPLOAD_FOLDER=images python main.py` we would rather not have the s3 operations dependant on user input including a slash. We could write our own input sanitation too but it's better to just use built-in libs to accomplish this.

I know that it was not a part of the challenge, but I think it's important to illustrate that any code you write could end up getting buried deep and misused.  This is why it's good to build in safeguards against misuse such as input sanitization, unit tests, and/or documentation. The end user, user (devs), operator (ops), and the author (you) of code is quite often all different people.

Reply:
That's how I had it originally, but changed it. Decided on the PurePath route this time.

Thanks for pointing this out, I will keep this in mind. I've noticed things can get buried deep pretty quickly.